nlp/model/augmented_lagrangian.py

```diff
@@ -128,6 +129,21 @@ def hess(self, *args, **kwargs):
         return self.hop(*args, **kwargs)
 
 
+    def magical_step(self, x, g):
+        """Compute a magical step for the slack variables.
+
+        This step minimizes the augmented Lagrangian with respect to the slack
+        variables only for a fixed set of decision variables. Here, g is the 
+        gradient computed by self.grad at x.
+        """
+        model = self.model
+        on = model.original_n
+        m_step = np.zeros(self.n)
+        m_step[on:] = -g[on:] / self.penalty
+        m_step = projected_step(x, m_step, model.Lvar, model.Uvar)
+        return m_step
+
+
```

Comment:
Actually, in this case, the magical step and the Newton step are the same because the objective is quadratic, as you outlined above. In general, they are not the same. 

I'm going to post my updated code to a different branch on my fork to keep it out of this pull request. We can discuss it further later.